tests/std/tests/P1222R4_flat_set/test.cpp

```diff
@@ -467,6 +553,131 @@ void test_extract_2() {
     assert_all_requirements_and_equals(fs, {}); // assert empty
 }
 
+void test_invariant_robustness() {
+    static int control_throw        = 0;
+    static constexpr int wont_throw = INT_MAX;
+
+    struct weird_elem {
+        static void countdown() {
+            if (control_throw == wont_throw) {
+                return;
+            }
+            if (--control_throw <= 0) {
+                throw 0; // will be caught by "catch (...)".
+            }
+        }
+
+        int key;
+
+        weird_elem(int k = 0) : key(k) {}
+        bool operator==(const weird_elem& other) const {
+            return key == other.key;
+        }
+
+        weird_elem(const weird_elem& other) : key(other.key) {
+            countdown();
+        }
+
+        weird_elem(weird_elem&& other) : key(exchange(other.key, 0)) {
+            countdown();
+        }
+
+        weird_elem& operator=(const weird_elem& other) {
+            countdown();
+            key = other.key;
+            return *this;
+        }
+
+        weird_elem& operator=(weird_elem&& other) {
+            countdown();
+            key = exchange(other.key, 0);
+            return *this;
+        }
+    };
+
+    class weird_container : public vector<weird_elem> {
+    private:
+        using base = vector<weird_elem>;
+
+    public:
+        using base::base;
+        weird_container(const weird_container&) = default;
+
+        // this copy-assigment cannot provide strong-guarantee for `this`:
+        weird_container& operator=(const weird_container&) = default;
+
+        // this move-ctor cannot provide strong-guarantee for `other`, and even successful, will leave elements of other
+        // in moved state:
+        weird_container(weird_container&& other) noexcept(false) {
+            reserve(other.size());
+            for (auto& e : other) {
+                push_back(std::move(e));
+            }
+        }
+
+        // this move-assignment cannot provide strong-guarantee for `this` and `other`, and even successful, will leave
+        // elements of other in moved state:
+        weird_container& operator=(weird_container&& other) noexcept(false) {
+            if (size() >= other.size()) {
+                for (size_t i = 0; i < other.size(); ++i) {
+                    at(i) = std::move(other.at(i));

```

Comment:
hmm, didnt notice this can be done with `std::move(other.begin(), other.end(), begin());`... will fix it later